### DISCUSSION TITLE
React 監聽 window.scroll 位置觸發動畫

### DISCUSSION THREAD

**Discussion by roroiii** [Lidemy/forum#248] [前端相關問題] [answered]
### **想請問使用 React 抓取 window.scroll ，例如當某個元件出現在螢幕畫面上就觸發事件，應該怎麼寫比較好。**

以前用 jQuery 寫監聽 window.scroll 是長這樣：
要觸發的元素出現在目前的畫面上就加上 is-fade 的 className，用新增的 className 改變 css 。

<img src="https://i.imgur.com/EAZPQ4f.gif" width="500"/>
（圖片是滑鼠往下滾動時）

```js
$(window)
    .scroll(function () {
      var windowHeight = $(window).height(),
        topWindow = $(window).scrollTop();
      $(".fadeUp:visible").each(function () {
        var targetPosition = $(this).offset().top;
        if (topWindow > targetPosition - windowHeight + 100) {
          $(this).addClass("is-fade");
        }
      });

      if ($(window).scrollTop() > 200) {
        $(".pageTop").fadeIn();
      } else {
        $(".pageTop").fadeOut();
      }
    })
    .trigger("scroll");
```

React 不能直接改動 DOM 元素，所以想說用 window.scroll 的位置來判斷觸發 useState()
已知 `window.scrollY` 可以得到目前滾動到哪裡的數值，所以可以用數值去明確判定大於多少之後觸發動畫
已知 `window.innerHeight` 可以得到目前視窗的高度。

```js
const [openText, setOpenText] = useState(false);
const prevScrollY = useRef(0);

useEffect(() => {
    const handleScroll = () => {
      const currentScrollY = window.scrollY;

      if (currentScrollY > 299) setOpenText(true);

      prevScrollY.current = currentScrollY;
    };
    window.addEventListener("scroll", handleScroll, { passive: true });
    return () => window.removeEventListener("scroll", handleScroll);
  }, []);
```
但是這樣如果畫面上有多個不同的 Component 與高度，在電腦與手機也因為 html 高度不一樣，就無法知道實際 window.scrollY 的數值是多少

想知道通常會怎麼處理這樣的情況？

有查過 react window.scroll 之類的關鍵字，但是大部分都是點擊按鈕回到最上方之類的功能，沒看到監聽位置觸發事件的寫法，已經找了滿多資料但是還是沒有眉目，所以才上來發問ＱＱ

**Comment by aszx87410** [comment, marked as answer]
其實你也可以用原本 jquery 那種用法，並沒有太大的不同
你原本 jq 拿的東西，在 react 絕對也都可以拿得到
或也可以換個方式直接去找一些 react 的 component，像這個：https://dbramwell.github.io/react-animate-on-scroll/
或這個：https://dev.to/chaituknag/animate-on-scroll-in-react-35e5

**Reply by roroiii** [reply]
其實我知道 jQuery 可以轉成為原生 JavaScript 只是一直寫不出來😂
也有去研究 react-spring 之類的套件，可能關鍵字沒下對就沒找到你提供的這個
感謝老師～～
